File: app/src/main/java/net/sourceforge/lifeograph/FragmentFilter.java
/* *********************************************************************************

 Copyright (C) 2012-2021 Ahmet Öztürk (aoz_2@yahoo.com)

 This file is part of Lifeograph.

 Lifeograph is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 Lifeograph is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Lifeograph.  If not, see <http://www.gnu.org/licenses/>.

 ***********************************************************************************/

package net.sourceforge.lifeograph;


import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;


public class FragmentFilter extends Fragment
{
    @Override
    public View onCreateView( LayoutInflater inflater,
                              ViewGroup container,
                              Bundle savedInstanceState ) {
        Log.d( Lifeograph.TAG, "FragmentFilter.onCreateView()" );

        ViewGroup rootView = ( ViewGroup ) inflater.inflate(
                R.layout.fragment_filter, container, false );

        // FILLING WIDGETS
        mEditSearch = rootView.findViewById( R.id.editTextSearch );
        mButtonSearchTextClear = rootView.findViewById( R.id.buttonClearText );

        mButtonShowTodoNot = rootView.findViewById( R.id.show_todo_not );
        mButtonShowTodoOpen = rootView.findViewById( R.id.show_todo_open );
        mButtonShowTodoProgressed = rootView.findViewById( R.id.show_todo_progressed );
        mButtonShowTodoDone = rootView.findViewById( R.id.show_todo_done );
        mButtonShowTodoCanceled = rootView.findViewById( R.id.show_todo_canceled );
        mSpinnerShowFavorite = rootView.findViewById( R.id.spinnerFavorites );

        if( Lifeograph.getScreenHeight() >= Lifeograph.MIN_HEIGHT_FOR_NO_EXTRACT_UI )
            mEditSearch.setImeOptions( EditorInfo.IME_FLAG_NO_EXTRACT_UI );

        // UI UPDATES (must come before listeners)
        updateFilterWidgets( Diary.diary.m_filter_active.get_status() );
        if( Diary.diary.is_search_active() ) {
            mEditSearch.setText( Diary.diary.get_search_text() );
            mButtonSearchTextClear.setVisibility( View.VISIBLE );
        }

        // LISTENERS
        mEditSearch.addTextChangedListener( new TextWatcher()
        {
            public void afterTextChanged( Editable s ) { }

            public void beforeTextChanged( CharSequence s, int start, int count, int after ) { }

            public void onTextChanged( CharSequence s, int start, int before, int count ) {
                if( mTextInitialized ) {
                    handleSearchTextChanged( s.toString() );
                    mButtonSearchTextClear.setVisibility(
                            s.length() > 0 ? View.VISIBLE : View.INVISIBLE );
                }
                else
                    mTextInitialized = true;
            }
        } );
        mButtonSearchTextClear.setOnClickListener( v -> mEditSearch.setText( "" ) );

        mButtonShowTodoNot.setOnClickListener( v -> handleFilterTodoChanged() );
        mButtonShowTodoOpen.setOnClickListener( v -> handleFilterTodoChanged() );
        mButtonShowTodoProgressed.setOnClickListener( v -> handleFilterTodoChanged() );
        mButtonShowTodoDone.setOnClickListener( v -> handleFilterTodoChanged() );
        mButtonShowTodoCanceled.setOnClickListener( v -> handleFilterTodoChanged() );

        mSpinnerShowFavorite.setOnItemSelectedListener( new AdapterView.OnItemSelectedListener()
        {
            public void onItemSelected( AdapterView< ? > pv, View v, int pos, long id ) {
                // onItemSelected() is fired unnecessarily during initialization, so:
                if( mInitialized )
                    handleFilterFavoriteChanged( pos );
                else
                    mInitialized = true;
            }

            public void onNothingSelected( AdapterView< ? > arg0 ) {
                Log.d( Lifeograph.TAG, "Filter Favorites onNothingSelected" );
            }

            private boolean mInitialized = false;
        } );

        Button buttonFilterReset = rootView.findViewById( R.id.buttonFilterReset );
        buttonFilterReset.setOnClickListener( v -> resetFilter() );
        Button buttonFilterSave = rootView.findViewById( R.id.buttonFilterSave );
        buttonFilterSave.setOnClickListener( v -> saveFilter() );

        return rootView;
    }

    @Override
    public void onAttach( @NonNull Context context ) {
        super.onAttach( context );

        Log.d( Lifeograph.TAG, "FragmentFilter.onAttach() - " + context.toString() );

        if( context instanceof FragmentEntryList.ListOperations )
            mListOperations = ( FragmentEntryList.ListOperations ) context;
        else
            throw new ClassCastException( context.toString() + " must implement ListOperations" );
    }

    @Override
    public void onResume() {
        super.onResume();

        mTextInitialized = false;
        mEditSearch.setText( Diary.diary.get_search_text() );
    }

    @Override
    public void onDetach() {
        super.onDetach();

        Log.d( Lifeograph.TAG, "FragmentFilter.onDetach() - " + this.toString() );

        mListOperations = null;
    }

    void handleSearchTextChanged( String text ) {
        Diary.diary.set_search_text( text.toLowerCase(), false );
        mListOperations.updateList();
    }

    void updateFilterWidgets( int fs ) {
        mButtonShowTodoNot.setChecked( ( fs & DiaryElement.ES_SHOW_NOT_TODO ) != 0 );
        mButtonShowTodoOpen.setChecked( ( fs & DiaryElement.ES_SHOW_TODO ) != 0 );
        mButtonShowTodoProgressed.setChecked( ( fs & DiaryElement.ES_SHOW_PROGRESSED ) != 0 );
        mButtonShowTodoDone.setChecked( ( fs & DiaryElement.ES_SHOW_DONE ) != 0 );
        mButtonShowTodoCanceled.setChecked( ( fs & DiaryElement.ES_SHOW_CANCELED ) != 0 );

        switch( fs & DiaryElement.ES_FILTER_FAVORED ) {
            case DiaryElement.ES_SHOW_FAVORED:
                mSpinnerShowFavorite.setSelection( 2 );
                break;
            case DiaryElement.ES_SHOW_NOT_FAVORED:
                mSpinnerShowFavorite.setSelection( 1 );
                break;
            case DiaryElement.ES_FILTER_FAVORED:
                mSpinnerShowFavorite.setSelection( 0 );
                break;
        }
    }

    void handleFilterTodoChanged() {
//        Diary.diary.m_filter_active.set_todo(
//                mButtonShowTodoNot.isChecked(),
//                mButtonShowTodoOpen.isChecked(),
//                mButtonShowTodoProgressed.isChecked(),
//                mButtonShowTodoDone.isChecked(),
//                mButtonShowTodoCanceled.isChecked() );
//
//        mListOperations.updateList();
    }

    void handleFilterFavoriteChanged( int i ) {
        boolean showFav = true;
        boolean showNotFav = true;

        switch( i ) {
            case 0:
                showFav = true;
                showNotFav = true;
                break;
            case 1:
                showFav = false;
                showNotFav = true;
                break;
            case 2:
                showFav = true;
                showNotFav = false;
                break;
        }

//        Diary.diary.m_filter_active.set_favorites( showFav, showNotFav );

        mListOperations.updateList();
    }

    void resetFilter() {
//        updateFilterWidgets( Diary.diary.m_filter_default.get_status() );
//        Diary.diary.m_filter_active.set_status_outstanding();
        mListOperations.updateList();
    }

    void saveFilter() {
        Lifeograph.showToast( R.string.filter_saved );
//        Diary.diary.m_filter_default.set( Diary.diary.m_filter_active );
    }

    private EditText mEditSearch = null;
    private Button mButtonSearchTextClear = null;
    private ToggleImageButton mButtonShowTodoNot = null;
    private ToggleImageButton mButtonShowTodoOpen = null;
    private ToggleImageButton mButtonShowTodoProgressed = null;
    private ToggleImageButton mButtonShowTodoDone = null;
    private ToggleImageButton mButtonShowTodoCanceled = null;
    private Spinner mSpinnerShowFavorite = null;

    private FragmentEntryList.ListOperations mListOperations;


    protected boolean mTextInitialized = false;
}

File: app/src/main/java/net/sourceforge/lifeograph/FragmentFilter.kt
/* *********************************************************************************

 Copyright (C) 2012-2025 Ahmet Öztürk (aoz_2@yahoo.com)

 This file is part of Lifeograph.

 Lifeograph is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 Lifeograph is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Lifeograph.  If not, see <http://www.gnu.org/licenses/>.

 ***********************************************************************************/

package net.sourceforge.lifeograph

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ImageButton
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton

class FragmentFilter : FragmentDiaryEditor(), RVAdapterFilterers.Listener  {
    // VARIABLES ===================================================================================
    override val            mLayoutId: Int = R.layout.fragment_filter
    override val            mMenuId: Int   = R.menu.menu_chart // same for now

    //protected val           mName: String

    private val             mElems: MutableList<Filterer> = ArrayList()
    private val             mSelectionStatuses: MutableList<Boolean> = ArrayList()
    private lateinit var    mAdapter: RVAdapterFilterers
    private lateinit var    mRecyclerView: RecyclerView
    private lateinit var    mFabAdd: FloatingActionButton
//    private lateinit var    mToolbar: HorizontalScrollView
    private lateinit var    mButtonReset : ImageButton
    private lateinit var    mButtonSave : ImageButton
    private lateinit var    mButtonRemove : ImageButton
    companion object {
        lateinit var mFilter: Filter
        lateinit var mStack: FiltererContainer
        fun setFilter(filter: Filter) {
            mFilter = filter
            mStack = mFilter._filterer_stack
        }
    }

    // METHODS =====================================================================================
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mRecyclerView = view.findViewById(R.id.list_elems)
        mAdapter = RVAdapterFilterers(mElems, mSelectionStatuses, this)
        mRecyclerView.adapter = mAdapter
        mRecyclerView.layoutManager = LinearLayoutManager(view.context)
        mFabAdd = view.findViewById(R.id.fab_add)
        mFabAdd.setOnClickListener { addFilterer() }
//        mToolbar = view.findViewById(R.id.toolbar_elem)
        mButtonReset = view.findViewById(R.id.reset_filter)
        mButtonSave = view.findViewById(R.id.save_filter)
        mButtonRemove = view.findViewById(R.id.remove_filterer)

        // LISTENERS
        mButtonReset.setOnClickListener { resetFilter() }
        mButtonSave.setOnClickListener { saveFilter() }
        mButtonRemove.setOnClickListener { removeFilterer() }

        // UI UPDATES (must come before listeners)
        updateFilterWidgets()
    }

    override fun onResume() {
        super.onResume()

        updateToolbarButtons()

        //Lifeograph.getActionBar().subtitle = mFilter._title_str
        Lifeograph.getActionBar().subtitle = mFilter._name // TODO implemet title str
    }

//    override fun onDestroy() {
//        super.onDestroy()
//    }
//
//    override fun updateMenuVisibilities() {
//        super.updateMenuVisibilities()
//    }

    // DiaryEditor interface methods
    override fun enableEditing() {
        super.enableEditing()

        updateToolbarButtons()
    }

    private fun updateToolbarButtons() {
        val flagEdit = Diary.getMain().is_in_edit_mode
        mButtonSave.visibility = if(flagEdit) View.VISIBLE else View.GONE
        mButtonReset.visibility = if(flagEdit) View.VISIBLE else View.GONE
        mButtonRemove.visibility = View.GONE
    }
    private fun updateFilterWidgets() {
        mElems.clear()
        for(filterer in mStack.m_pipeline)
            mElems.add(filterer)
    }

    private fun addFilterer() {
        DialogPicker(requireContext(), object: DialogPicker.Listener{
            override fun onItemClick(item: RViewAdapterBasic.Item) {
                when(item.mId) {
                    "STATUS" -> { mStack.add_filterer_status( DiaryElement.ES_FILTER_TRASHED ) }
                    "FAVORITE" -> { mStack.add_filterer_favorite( true ) }
                    "TRASHED" -> { mStack.add_filterer_trashed( true ) }
                    "IS" -> { mStack.add_filterer_is( 404 /*TODO: DEID.UNSET*/, true ) }
                    "HASTAG" -> { mStack.add_filterer_tagged_by( null, true ) }
                }
                mSelectionStatuses.add(false)
                updateFilterWidgets()
                mAdapter.notifyItemChanged( mStack.m_pipeline.size - 1 )
            }

            override fun populateItems(list: RVBasicList) {
                list.clear()

                list.add(RViewAdapterBasic.Item(
                    Lifeograph.getStr(R.string.filter_status), "STATUS",
                    R.drawable.ic_filter))
                list.add(RViewAdapterBasic.Item(
                    Lifeograph.getStr(R.string.filter_favorite), "FAVORITE",
                    R.drawable.ic_filter))
                list.add(RViewAdapterBasic.Item(
                    Lifeograph.getStr(R.string.filter_trashed), "TRASHED",
                    R.drawable.ic_filter))
                list.add(RViewAdapterBasic.Item(
                    Lifeograph.getStr(R.string.filter_is), "IS",
                    R.drawable.ic_filter))
                list.add(RViewAdapterBasic.Item(
                    Lifeograph.getStr(R.string.filterer_has_tag), "HASTAG",
                    R.drawable.ic_filter))
            }
        }).show()
    }
    private fun removeFilterer() {
        val itemsToRemove = mutableListOf<Filterer>()
        val indicesToRemove = mutableListOf<Int>()

        // First, identify all items and their original indices to be removed
        // Iterate backwards to safely use indices for removal from mSelectionStatuses
        for (i in mSelectionStatuses.indices.reversed()) {
            if (mSelectionStatuses[i]) {
                // Check if the index is valid for mElems before trying to access it
                if (i < mElems.size) {
                    itemsToRemove.add(mElems[i]) // Add the Filterer object
                    indicesToRemove.add(i)       // Add the original index in mElems
                } else {
                    // Log an error or handle inconsistency if mSelectionStatuses is out of sync with mElems
                    Log.e("FragmentFilter", "Inconsistency: Selection status for out-of-bounds index $i")
                }
            }
        }

        if (itemsToRemove.isEmpty()) {
            // No items were selected for removal
            return
        }

        // 1. Remove from the primary data source (mStack)
        for (filtererToRemove in itemsToRemove) {
            mStack.remove_filterer(filtererToRemove)
        }

        // 2. Remove from the local lists used by the adapter (mElems and mSelectionStatuses)
        //    It's crucial to remove from the highest index to lowest to avoid shifting issues
        //    if you were modifying mElems and mSelectionStatuses directly in the first loop.
        //    Since indicesToRemove is already sorted from high to low (due to reversed iteration),
        //    this is straightforward.
        for (index in indicesToRemove) {
            if (index < mElems.size) { // Double-check bounds, though should be fine if synced
                mElems.removeAt(index)
            }
            if (index < mSelectionStatuses.size) { // Also for selection statuses
                mSelectionStatuses.removeAt(index)
            }
        }

        // 3. Notify the adapter about the removals.
        //    This is trickier for multiple, non-contiguous removals if you want animations.
        //    The most straightforward way (though less performant for animations)
        //    is to rebuild mElems and call notifyDataSetChanged or use ListAdapter.
        //    For individual notifications, you'd call notifyItemRemoved for each index
        //    (adjusting for previous removals) or group contiguous removals.

        // Option A: Simpler, but loses animations and less efficient
        // updateFilterWidgets() // Repopulate mElems from the updated mStack
        // mAdapter.notifyDataSetChanged()

        // Option B: More granular notifications (if you keep mElems and mSelectionStatuses manually updated)
        // This assumes `indicesToRemove` contains the original indices sorted from highest to lowest.
        // As you call notifyItemRemoved, subsequent items' indices effectively shift.
        // However, since RecyclerView processes these removals, it can often handle it.
        // But it's safer to notify in reverse order of removal as well.
        for (index in indicesToRemove.sortedDescending()) { // Ensure they are sorted high to low
            mAdapter.notifyItemRemoved(index)
        }
        // After removals, you might need to notify for range changes in the remaining items
        // to update their positions if your adapter relies on absolute positions for binding.
        // A simple way if many items might have shifted:
        if (indicesToRemove.isNotEmpty()) {
            // Find the lowest index that was removed
            val lowestRemovedIndex = indicesToRemove.minOrNull() ?: 0
            if (lowestRemovedIndex < mElems.size) {
                mAdapter.notifyItemRangeChanged(lowestRemovedIndex, mElems.size - lowestRemovedIndex)
            } else if (mElems.isNotEmpty()){ // If all items up to the end were removed, but some remain
                mAdapter.notifyItemRangeChanged(0, mElems.size)
            }
            //else if (indicesToRemove.isNotEmpty()) {
                // All items were removed, no range change needed, notifyDataSetChanged would also work
                // or ensure the last notifyItemRemoved covered the last item.
            //}
        }


        // 4. Update UI elements
        updateActionBarSubtitle()
        if (mAdapter.mSelCount == 0 || mSelectionStatuses.none { it }) { // Ensure selection is truly cleared
            exitSelectionMode()
        }
    }

    private fun saveFilter() {
        mFilter._definition = mStack._as_string

        Diary.getMain().updateAllEntriesFilterStatus()
    }
    private fun resetFilter() {
        mStack = mFilter._filterer_stack
        updateToolbarButtons()
        updateFilterWidgets()
    }

    override fun updateActionBarSubtitle() {
        val selCount = mAdapter.mSelCount
        val itmCount = mAdapter.itemCount
        Lifeograph.getActionBar().subtitle =
            if( selCount > 0 ) "${mFilter._name} ($selCount/$itmCount)"
            else               "${mFilter._name} ($itmCount)"
    }

    override fun enterSelectionMode(): Boolean {
        return if(Diary.getMain().is_in_edit_mode) {
            mButtonRemove.visibility = View.VISIBLE
            true
        }
        else false
    }

    override fun exitSelectionMode() {
        updateActionBarSubtitle()
        mButtonRemove.visibility = View.GONE
    }
}
